models/reporter.go

```diff
@@ -622,26 +474,46 @@ func (r *Reporter) SelectFilters(ctx context.Context, periodStartFromSec, period
 	return &result, nil
 }
 
-func (r *Reporter) queryFilters(ctx context.Context, periodStartFromSec, periodStartToSec int64, mainMetricName, query string) ([]*customLabel, float32, error) {
+func (r *Reporter) queryFilters(ctx context.Context, periodStartFromSec,
+	periodStartToSec int64, dimensionName, mainMetricName string, tmplQueryFilter *template.Template, queryDimensions, queryLabels map[string][]string) ([]*customLabel, float32, error) {
 	durationSec := periodStartToSec - periodStartFromSec
 	var labels []*customLabel
-	rows, err := r.db.QueryContext(ctx, fmt.Sprintf(query, mainMetricName), periodStartFromSec, periodStartToSec)
+
+	tmplArgs := struct {
+		MainMetric    string
+		DimensionName string
+		Dimensions    map[string][]string
+		Labels        map[string][]string
+	}{
+		mainMetricName,
+		dimensionName,
+		queryDimensions,
+		queryLabels,
+	}
+
+	var queryBuffer bytes.Buffer
+
+	if err := tmplQueryFilter.Execute(&queryBuffer, tmplArgs); err != nil {
+		return nil, 0, errors.Wrapf(err, "cannot execute tmplQueryFilter %s", queryBuffer.String())
```

Comment:
https://golang.org/pkg/fmt/#hdr-Printing
> If an operand implements method String() string, that method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).

You don't have to call `String()` explicitly.

Reply:
Yes, thank you, I saw - but tests throw warnings

```
go test -v ./...
# github.com/percona/qan-api2/models
models/reporter.go:497:18: Wrapf format %s has arg queryBuffer of wrong type bytes.Buffer
models/reporter.go:502:18: Wrapf format %s has arg queryBuffer of wrong type bytes.Buffer
models/reporter.go:510:19: Wrapf format %s has arg queryBuffer of wrong type bytes.Buffer
models/reporter.go:516:18: Wrapf format %s has arg queryBuffer of wrong type bytes.Buffer
models/reporter.go:526:20: Wrapf format %s has arg queryBuffer of wrong type bytes.Buffer
models/reporter.go:531:19: Wrapf format %s has arg queryBuffer of wrong type bytes.Buffer
```